Mage.Sets/src/mage/cards/r/RiverSongsDiary.java

```diff
@@ -0,0 +1,186 @@
+        super(ability);
+    }
+
+    @Override
+    public RiverSongsDiaryTriggeredAbility copy() {
+        return new RiverSongsDiaryTriggeredAbility(this);
+    }
+
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.SPELL_CAST;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        Spell spell = game.getStack().getSpell(event.getTargetId());
+        System.out.println(spell.getFromZone() + "fromZone");
+        if (spell == null || !spell.isInstantOrSorcery(game) || !spell.getFromZone().equals(Zone.HAND)) {
+            return false;
+        }
+        this.getEffects().clear();
+        this.addEffect(new RiverSongsDiaryExileEffect(spell, game));
+        return true;
+    }
+
+    @Override
+    public String getRule() {
+        return "Whenever a player casts an instant or sorcery spell from their hand, " +
+                "exile it instead of putting it into a graveyard as it resolves.";
+    }
+}
+
+class RiverSongsDiaryExileEffect extends ReplacementEffectImpl {
+
+    private final MageObjectReference mor;
+
+    RiverSongsDiaryExileEffect(Spell spell, Game game) {
+        super(Duration.WhileOnStack, Outcome.Benefit);
+        this.mor = new MageObjectReference(spell, game);
+    }
+
+    private RiverSongsDiaryExileEffect(final RiverSongsDiaryExileEffect effect) {
+        super(effect);
+        this.mor = effect.mor;
+    }
+
+    @Override
+    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+        Spell sourceSpell = game.getStack().getSpell(event.getTargetId());
+        if (sourceSpell == null || sourceSpell.isCopy()) {
+            return false;
+        }
+        Player player = game.getPlayer(sourceSpell.getOwnerId());
+        if (player == null) {
+            return false;
+        }
+        player.moveCardsToExile(
+                sourceSpell, source, game, false,
+                CardUtil.getExileZoneId(game, source),
+                CardUtil.getSourceName(game, source)
+        );
+        return true;
+    }
+
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE;
+    }
+
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+        ZoneChangeEvent zEvent = ((ZoneChangeEvent) event);
+        if (zEvent.getFromZone() != Zone.STACK
+                || zEvent.getToZone() != Zone.GRAVEYARD
+                || event.getSourceId() == null
+                || !event.getSourceId().equals(event.getTargetId())
+                || mor.getZoneChangeCounter() != game.getState().getZoneChangeCounter(event.getSourceId())) {
+            return false;
+        }
+        Spell spell = game.getStack().getSpell(mor.getSourceId());
+        return spell != null && spell.isInstantOrSorcery(game) && spell.getFromZone().equals(Zone.HAND);
```

Comment:
what's the point of repeating these checks? from zone and instant/sorcery aren't going to change, and were checked by the trigger